src/main/java/picard/sam/BamErrorMetric/CollectBamErrorMetrics.java

```diff
@@ -0,0 +1,470 @@
+                    VCF == null ? Collections.emptyIterator() : new VCFFileReader(VCF, true).iterator())
+        ) {
+
+            final SAMSequenceDictionary sequenceDictionary = referenceSequenceFileWalker.getSequenceDictionary();
+            if (sam.getFileHeader().getSortOrder() != SAMFileHeader.SortOrder.coordinate)
+                throw new PicardException("Input BAM must be sorted by coordinate");
+
+            sequenceDictionary.assertSameDictionary(sam.getFileHeader().getSequenceDictionary());
+
+            final IntervalList regionOfInterest = getIntervals(sequenceDictionary);
+
+            log.info("Getting SamLocusIterator");
+
+            final SamLocusIterator samLocusIterator = new SamLocusIterator(sam, regionOfInterest);
+
+            // Now go through and compare information at each of these sites
+            samLocusIterator.setEmitUncoveredLoci(false);
+            samLocusIterator.setMappingQualityScoreCutoff(MIN_MAPPING_Q);
+            samLocusIterator.setQualityScoreCutoff(MIN_BASE_Q);
+
+            log.info("Using " + aggregatorList.size() + " aggregators.");
+
+            aggregatorList.forEach(la ->
+                    IOUtil.assertFileIsWritable(new File(OUTPUT + la.getSuffix())));
+
+            // iterate over loci
+            log.info("Starting iteration over loci");
+
+            final SamLocusAndReferenceIterator iterator = new SamLocusAndReferenceIterator(referenceSequenceFileWalker, samLocusIterator);
+
+            // This hasNext() call has side-effects. It loads up the index and makes sure that the iterator is really ready for
+            // action. Calling this allows for the logging to be more accurate.
+            iterator.hasNext();
+            log.info("Really starting iteration now.");
+
+            for (final SAMLocusAndReference info : iterator) {
+                if (random.nextDouble() > PROBABILITY)
+                    continue;
+
+                nTotalLoci++;
+
+                // while there is a next (non-filtered) variant and it is before the locus, advance the pointer.
+                if (advanceIteratorAndCheckLocus(vcfIterator, info.getLocus(), sequenceDictionary)) {
+                    log.debug(String.format("Skipping locus from VCF: %s", vcfIterator.peek().toStringWithoutGenotypes()));
+                    nSkippedLoci++;
+                    continue;
+                }
+
+                addLocusBases(aggregatorList, info);
+
+                nProcessedLoci++;
+                progressLogger.record(info.getLocus().getSequenceName(), info.getLocus().getPosition());
+
+                if (MAX_LOCI != 0 && nProcessedLoci >= MAX_LOCI) {
+                    log.warn("Early stopping due to having processed MAX_LOCI loci.");
+                    break;
+                }
+            }
+
+        } catch (IOException e) {
+            log.error("A problem occurred:", e.getMessage());
```

Comment:
In general you should always include the exception when logging so a stack crawl is printed. Once you do that you no longer need to get the exception message as it's already included in the output. E.g. `log.error("A problem occurred", e);`